src/main/java/com/icthh/xm/tmf/ms/activation/repository/SagaTransactionRepository.java

```diff
@@ -7,14 +7,25 @@
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 
 public interface SagaTransactionRepository extends JpaRepository<SagaTransaction, String> {
 
     Page<SagaTransaction> findAllBySagaTransactionState(SagaTransactionState sagaTransactionState, Pageable pageable);
 
     long countByCreateDateBeforeAndSagaTransactionState(Instant date, SagaTransactionState state);
 
+    long countByCreateDateBeforeAndSagaTransactionStateAndTypeKey(Instant date, SagaTransactionState state, String typeKey);
+
     Optional<SagaTransaction> findByKey(String key);
 
     Optional<SagaTransaction> findOneById(String id);
+
+    @Query(value = "select COUNT(e.status)" +
+        " from saga_transaction t, saga_event e" +
+        " where t.type_key = ?1" +
+        " AND t.id = e.transaction_id" +
+        " AND e.status = ?2" +
+        " AND e.create_date < ?3", nativeQuery = true)
+    long countByTypeKeyAndEventStatusAndCreatedDateBefore(String typeKey, String eventStatus, Instant date);
```

Comment:
Why native sql is used? Is it possible to use criteria or at least JPQL?

Reply:
> Why native sql is used? Is it possible to use criteria or at least JPQL?

I can't use JPQL or criteria builder because this query use joining of two tables (saga_transaction + saga_event). But at activation-ms we don't have relations between SagaTransaction entity and SagaEvent.

Do you suggest to add relations between SagaTransaction and SagaEvent?

Reply:
@pkev777 @vkirichenko 
there is a possibility to join unassociated entities:
https://www.youtube.com/watch?v=wBD66qYcKuU

but we need to be aware of the performance. So you can try.

Reply:
> @pkev777 @vkirichenko there is a possibility to join unassociated entities: https://www.youtube.com/watch?v=wBD66qYcKuU
> 
> but we need to be aware of the performance. So you can try.

hm, indeed. I will try, thanks